sklearn/decomposition/_pca.py

```diff
@@ -533,33 +562,77 @@ def _fit_full(self, X, n_components):
+            X_centered = xp.asarray(X, copy=True) if self.copy else X
+            X_centered -= self.mean_
+            x_is_centered = not self.copy
+
+            if not is_array_api_compliant:
+                # Use scipy.linalg with NumPy/SciPy inputs for the sake of not
+                # introducing unanticipated behavior changes. In the long run we
+                # could instead decide to always use xp.linalg.svd for all inputs,
+                # but that would make this code rely on numpy's SVD instead of
+                # scipy's. It's not 100% clear whether they use the same LAPACK
+                # solver by default though (assuming both are built against the
+                # same BLAS).
+                U, S, Vt = linalg.svd(X_centered, full_matrices=False)
+            else:
+                U, S, Vt = xp.linalg.svd(X_centered, full_matrices=False)
+            explained_variance_ = (S**2) / (n_samples - 1)
+
         else:
-            U, S, Vt = xp.linalg.svd(X, full_matrices=False)
+            assert self._fit_svd_solver == "covariance_eigh"
+            # In the following, we center the covariance matrix C a-posteriori
+            # (without centering the data X first) to avoid an unecessary copy
+            # of X. Note that the mean_ attribute is still needed to center
+            # test data in the transform method.
+            #
+            # Note: at the time of writing, `xp.cov` does not exist in the
+            # Array API standard:
+            # https://github.com/data-apis/array-api/issues/43
+            #
+            # Besides, using `numpy.cov`, as of numpy 1.26.0, would not be
+            # memory efficient for our use case when `n_samples >> n_features`:
+            # `numpy.cov` centers a copy of the data before computing the
+            # matrix product instead of substracting a small `(n_features,
+            # n_features)` square matrix, a posteriori, as we do below.
+            x_is_centered = False
+            C = X.T @ X
+            C -= (
+                n_samples
+                * xp.reshape(self.mean_, (-1, 1))
+                * xp.reshape(self.mean_, (1, -1))
+            )
+            C /= n_samples - 1
+            eigenvals, Eigenvecs = xp.linalg.eigh(C)
+            eigenvals = xp.flip(eigenvals, axis=0)
+            Eigenvecs = xp.flip(Eigenvecs, axis=1)
+
+            # The covariance matrix C is positive semi-definite by
+            # construction. However, the eigenvalues returned by xp.linalg.eigh
+            # can be slightly negative due to numerical errors. This would be
+            # an issue for the subsequent sqrt, hence the manual clipping.
+            eigenvals[eigenvals < 0.0] = 0.0
+            explained_variance_ = eigenvals
+
+            # Re-construct synthetic SVD components to be consistent with the
+            # other solvers.
+            S = xp.sqrt(eigenvals * (n_samples - 1))
+            Vt = Eigenvecs.T
```

Comment:
```suggestion
            Eigenvecs = xp.flip(Eigenvecs, axis=1)

            # The covariance matrix C is positive semi-definite by
            # construction. However, the eigenvalues returned by xp.linalg.eigh
            # can be slightly negative due to numerical errors. This would be
            # an issue for the subsequent sqrt, hence the manual clipping.
            eigenvals[eigenvals < 0.0] = 0.0
            explained_variance_ = eigenvals

            # Re-construct synthetic SVD components to be consistent with the
            # other solvers.
            S = xp.sqrt(eigenvals * (n_samples - 1))
            Vt = eigenvecs.T
```